Auto install script for docker/dockur/winapps with 2 network options (DHCP LAN + macvlan & host network)

itiligent opened the pull request:
### **This script offers two Docker deployment options for running WinApps on Docker:**

#### **Option 1**: [Easy] Deploys a Windows container with Dockur's default network settings 
  - A fully automated Docker/Dockur/Windows build with just a few simple prompts
  - The Windows container shares an IP with the host
     - (Optionally) restrict incoming RDP & VNC access to 127.0.0.1


#### 2. **Option 2**: [Advanced] Deploys a Windows container with macvlan networking 
   - Macvlans allow a container to act as a full LAN client with a separate DHCP IP & MAC address
      - The script walks the user through the required CIDR inputs & automatically builds all routing schema

(Note: Script will auto install flatpak version of FreeRDP3 if a distro version is not available)

### Script Instructions
 1. Run the script (without root permissions) & answer the script prompts to customize your build
 2. Follow the Windows install via VNC http://127.0.0.1:8006 or http://x.x.x.x:8006
 3. Restart Windows via the VNC console when the installation is complete
 4. Refresh Docker group permissions by running `newgrp docker`
 5. Test rdp from Linux by running `~/test-rdp.sh` (this test script is created at install to suit the individual build settings)
 6. Install your preferred Windows applications as needed via RDP or VNC
 7. Run  `~/winapps/installer.sh` to install WinApps

      
#### More about macvlans:
   - Macvlan pros:
     - Allows 3-way sharing between the container, local host, and LAN
     - Provides direct communication between the container and NIC, (no NAT or port forwarding)
     - DHCP makes IP management consistent, aiding firewall rules and network policy
     - Some applications may require unique IP/MAC addresses for licensing
     - (Optionally) restrict incoming RDP & VNC access to 127.0.0.1
     - Automates a complex build into a repeatable task with potential for rapid & low cost DR virtual desktop infrastructure 
   - Macvlan cons:
    - As macvlans use multiple MAC addresses per interface:
       - This option is available only for ethernet-connected systems
       - Macvlans are NOT compatible with WiFi NICs

The script is curently Debian/Ubuntu focused but can be used as a framework for other distros

itiligent commented:
I made a final change to add a sound on/off prompt menu and simplified install instructions as a reboot seems best to avoid group permissions not always refreshing after docker is added via one continuous script.

oskardotglobal commented:
I don't know about this one. 
The script makes a lot of assumptions and there is no documentation for it. For this to be merged the according installation documentation would have to be updated (but only for Debian-based distros).

itiligent commented:
Firstly, this is a great project which is why I'd like to help it towards user friendliness & made more accessible for less technical folk. 

Regarding documentation, the script is a actually a coalescence of the very minimal amount of installation documentation I've been able to uncover. Another view (and a key part of my intent for submitting it) might be that reverse engineering it (its all  quite clearly commented) would help produce a much more comprehensive set of real world install instrutions than exists anywere because in writing it I uncovered many gaps in the current docs that had to be worked though.  

Could you please explain what documetation would be needed? Perhaps I could help?   At a minimum this scipt could perhpas be viewed as the quantum of actual verified install steps to build the WinApps recommended enviroment?

Regarding script assumptions, I dont quite understanfd what you mean there. There are virtually no assumprions contained (as far as I can tell) and the user is guided through every key decision point I've uncovered in my own use of WinApps.. Could you please give some examples of assumptions and perhaps I could address those? 

Regarding Debian based:   perhaps this is what you mean by assumptions?  I've put this install flow together as a kind of orderly inatall approach with a view that it could be expanded to other mainstream distros in the same way as I've smoothed out Docker's repo awkwardess od install between Debian distros.  

Happy to help!

oskardotglobal commented on winapps-for-dockur-setup.sh:
```diff
@@ -0,0 +1,780 @@
+    echo " Select a Windows installer source:"
+    echo "    1) Use your own ISO: ${DEFAULT_INSTALL_ISO}"
+    echo "    2) Provide a new ISO path"
+    echo "    3) No ISO (script will download Windows)"
```
This should be the default.

itiligent replied:
keeping the default compose in full is a good idea for continuity of docs. The main reason I cut it down was to make it easier to work with and review. I notice in the past few days the winapps.config deault file has been improved, so this should also be made to match.

An the matter of sound support, I would really prefer that the sound on/off hack on line 586 to not be needed, and instead all just be toggled on/off via  flags in the winapps.conf file as per the FreeRD3 docs suggest  - this would also make user sound permanent as this hack will get overwritten as soon as /upgrade/re-clone event takes place.

oskardotglobal commented on winapps-for-dockur-setup.sh:
```diff
@@ -0,0 +1,780 @@
+      - data:/storage
+      - ${HOME}:/shared
+      - ./oem:/oem
+      - $INSTALL_ISO:/custom.iso
```
Pretty sure this doesn't work even with your above empty image, just don't add the line if INSTALL_ISO is empty

itiligent replied:
it works on all flavours of debian. U22 & U24 dont like a blank entry there - at line 643 there is a workaround that adds a dummy file there if no ISO is configured. This was a far simpler and self contained soluton rather than adding a conditional override and extra malarkey to achieve in a script that may need to support extra distros - but I'm open to suggestions to improve my bash! 

The above works because if there is no *working* ISO image in the path that is specified, Dockur just defaults back to the DEFAULT_VERSION value and installs Windows from a download.  It get tricky when you are trying to straddle and support two independent pieces of software, so I'm trying to keep it as simple as possible.

oskardotglobal replied:
Yeah, that was what I was talking about

oskardotglobal commented on winapps-for-dockur-setup.sh:
```diff
@@ -0,0 +1,780 @@
+############################################################################################################
+
+# Container defaults
+DEFAULT_VERSION="win10" # see https://github.com/dockur/windows for all options
```
This should be either `win11` or `tiny11`.

itiligent replied:
No problem, thats in there just coz I actaully use a win10 distribution point as testing with downloads takes too long.

oskardotglobal reviewed (changes requested):
I've given the script another review and I missed that it clones winapps before. Just some minor things that need addressing.

Also, I would really prefer if you overwrote the default compose.yaml instead of writing a new one since that would preserve all the comments we added to it. We could also use env vars with fallback values inside the compose file so you could search and replace them.

The other major thing missing is that this script has to be mentioned within the README.

LDprg commented:
I feel like we shouldn't merge an debian only install script.

It should at least support debian, arch and fedora distros. This shouldn't be too much of a problem to implement and still avoid alot of confusion.

itiligent commented:
Thanks for the feedback. Take your time and play with it and get a feel for it and what I've tried to achieve through it. More eyes will always help find the things I mght not have thought of.

itiligent commented:
Ok so I've done a bunch of work on this today.  

- I've fixed the blank ISO issue, (it urns out that the custom ISO path must be specified below all volumes in the compose file else windows  fails to find an image after 2nd boot)
- Added lots of validation to the ISO inputs to check to make sure that an ISO file also exisits before the scipt will accept it
-  Default docker compose template files have been re-added, 
- New winapps.conf template is incorporated.
-  Added ability to override the auto IP adressing for Macvlan (offers flexibility for cases where several containers might run a single host or there are multiple NICS and a need for more than a  /30)  
- Lots of spell checks and grammar tidy ups
-  I've also changed the colours to match the WinApps logo. 
 
My preference is to gather all the feedback on this debian version, and once everyone is happy with it, then I can look to expanding it to Fedora and Arch

oskardotglobal commented:
Thanks for the great work, @itiligent. We'll properly review this once you're ready

itiligent commented:
Closing as I have progressed a much more refined install script version, and I will re-submit again soon.
